cirq-google/cirq_google/calibration/workflow.py

```diff
@@ -268,13 +270,33 @@ def prepare_characterization_for_moments(
         operations matched by gates_translator, or it mixes a single qubit and two qubit gates.
     """
     if initial is None:
-        allocations: List[Optional[int]] = []
-        calibrations: List[RequestT] = []
-        pairs_map: Dict[Tuple[Tuple[Qid, Qid], ...], int] = {}
+        initial = []
+
+    if isinstance(circuits, Circuit):
+        return _prepare_characterization_for_circuit_moments(
+            circuits, options, gates_translator, merge_subsets, initial
+        )
     else:
-        allocations = []
-        calibrations = list(initial)
-        pairs_map = {calibration.pairs: index for index, calibration in enumerate(calibrations)}
+        circuits_with_calibration = []
+        requests = list(initial)
+        for circuit in circuits:
+            circuit_with_calibration, requests = _prepare_characterization_for_circuit_moments(
+                circuit, options, gates_translator, merge_subsets, requests
+            )
+            circuits_with_calibration.append(circuit_with_calibration)
+        return circuits_with_calibration, requests
```

Comment:
It's really useful to have this logic/boilerplate somewhere since I've found myself writing it a lot. 

That being said: I'm really nervous about having a different return type based on the input type. This is usually a pretty bad thing. Have you considered a) always returning a list of circuits (maybe of length 1) or b) keeping the old version of the function that works one circuit at a time and introducing a wrapper that's like `prepare_characterizations_for_circuits` which always returns a list (it can optionally handle one circuit input as a special case). This is basically how you've structured the PR already but keeping the name of the old function (and keeping it public) and adding a new API

Reply:
I wanted to reserve `prepare_characterizations_for_circuits` for future calibrations which are not only moment-aware but consider the entire circuit structure.... There is `prepare_characterizations_for_moments` and `prepare_characterizations_for_operations` now, we might need `prepare_characterizations_for_circuits`. Or maybe this naming is a little bit broken.

What is really missing is the function overloading which doesn't exist in Python. I would argue this code is fine exactly because of this deficiency of a language? Balint, Matthew, what do you think?

Reply:
Actually, what would be even better is to support arbitrary nested sequences. I'm playing right now with some code I have 2 dimensional array already; i.e., have CIRCUIT_TREE analogous to the OP_TREE. I see that typing for OP_TREE is non-trivial, maybe this should be a follow up by someone in Cirq team?

Reply:
Even with languages with overloading, the functions have consistent return types, which is my issue with the current proposed change

Reply:
You're right. I named the new method `prepare_characterizations_for_circuits_moments`, how does it sound?